### PR TITLE
Don't offer import completions in non-module files unless "--module" is set

### LINKED ISSUE #10178: Consider allowing access to UMD globals from modules
Feedback from #7125 is that some people actually do mix and match global and imported UMD libraries, which we didn't consider to be a likely scenario when implementing the feature.

Three plausible options:
1. Do what we do today - bad because there's no good workaround
2. Allow some syntax or configuration to say "this UMD global is actually available globally" - somewhat complex, but doable
3. Allow access to all UMD globals regardless of context - misses errors where people forget to import the UMD global in a file. These are presumably somewhat rare, but it would be dumb to miss them

Sounds like it would work but probably wouldn't:
1. Flag imported UMD modules as "not available for global" - bad because UMD modules will be imported in declaration files during module augmentation. It'd be weird to have differing behavior of imports from implementation files vs declaration files.

I'm inclined toward option 3 for simplicity's sake, but could see option 2 if there's some reasonably good syntax or configuration we could use in a logical place. Detecting the use of a UMD global in a TSLint rule would be straightforward if someone wanted to do this.

One path forward would be to implement option 3 and if it turns out people make the "forgot to import" error often, add a tsconfig option `globals: []` that explicitly specifies which UMD globals are allowed.

### LINKED ISSUE #22750: Auto-imports should not be offered in non-module files
**Code**
```
> npm init -y
> tsc --init
> npm install @types/jquery
> npm install @types/express
> code a.ts
```
Type
```
$.ajax();
ss
```
**Actual behavior:**

image

**Expected behavior:**

We shouldn't offer auto imports in a global file (i.e. until there's at least one other import). This was really confusing for me - I was working in the TS codebase, selected a function name that looked like what I wanted, then got a screenfull of compiler errors when I tried to build because I had inadvertently changed the containing file into a module (thus breaking everything). Because the import appeared off-screen at the top of the file I had no idea what even happened until I ran `git diff`.

### PR DESCRIPTION
**Pull request by andy-ms**
Fixes #22750

Note: does not include the logic for looking through a source file for a UMD use. Personally, I think we should just fixx #10178 (intentional misspelling to prevent GH from closing that) so that using modules doesn't break your existing code!

### CODE REVIEW

File: src/services/completions.ts [hunk 1]
```diff
@@ -1169,13 +1169,17 @@ namespace ts.Completions {
                 }
             }
 
-            // Don't suggest import completions for a commonjs-only module
-            if (preferences.includeCompletionsForModuleExports && !(sourceFile.commonJsModuleIndicator && !sourceFile.externalModuleIndicator)) {
```
Comment (mhegazy) [inline src/services/completions.ts]: i am a bit confused now. why was not that working already? in the issue the file is not a module (i.e. no externalModuleIndicator and no commonJsModuleIndicator), so we should not have included the completions, right?

Comment (andy-ms) [reply src/services/completions.ts]: I was thinking that even in a currently import-less file we should still add import completions if the user has `--module` set in their tsconfig. (see https://github.com/Microsoft/TypeScript/issues/22750#issuecomment-374799396)

Comment (andy-ms) [reply src/services/completions.ts]: The reason it wasn't working already is we would always add import completions for a file with no commonjs imports/exports.

Comment (mhegazy) [reply src/services/completions.ts]: I thought that is what https://github.com/Microsoft/TypeScript/pull/22880 was doing.. but looking at it again, the condition does not seem right.. the problem seems to be more complicated than i originally thought..

how about somethign like:

```ts
function shouldOfferImportCompletions(): boolean {
                if (preferences.includeCompletionsForModuleExports) {
                    // sure to have a module
                    if (sourceFile.externalModuleIndicator) return true;

                    // sure to have a .js file with a require, do not mess with that file
                    if (sourceFile.commonJsModuleIndicator) return false;

                    // either this file is not really a module, or it is a new file added to the project
                    // see if there are other "source" files that we can infer from
                    if (find(program.getSourceFiles(), s => !s.isDeclarationFile && !program.isSourceFileFromExternalLibrary(s) && !!s.externalModuleIndicator))
                        return true;

                    // either this is an empty project, or one with no modules in it
                    // for .js files we should fall on the safe side -- i think :)
                    if (isInJavaScriptFile(sourceFile)) return false;

                    // in the absense of any other info, then may be we can find something in the compiler options
                    if (compilerOptions.module !== undefined) return compilerOptions.module > ModuleKind.None;
                    if (compilerOptions.target >= ScriptTarget.ES2016) return true;
                    if (compilerOptions.noEmit) return true;
                }
                return false;
            }
```

**Comment by andy-ms** [comment]
@DanielRosenwasser Open a new issue if you want us to do something about UMD globals here.

**Pull request merged**